core/src/main/java/org/apache/calcite/rel/rules/SortRemoveConstantKeys.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to you under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.rel.rules;
+
+import org.apache.calcite.plan.RelOptPredicateList;
+import org.apache.calcite.plan.RelOptRule;
+import org.apache.calcite.plan.RelOptRuleCall;
+import org.apache.calcite.rel.RelCollationTraitDef;
+import org.apache.calcite.rel.RelCollations;
+import org.apache.calcite.rel.RelFieldCollation;
+import org.apache.calcite.rel.core.Sort;
+import org.apache.calcite.rel.metadata.RelMetadataQuery;
+import org.apache.calcite.rex.RexInputRef;
+import org.apache.calcite.rex.RexNode;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Planner rule that removes
+ * a {@link org.apache.calcite.rel.core.Sort} if its input is already sorted.
+ *
+ * <p>Requires {@link RelCollationTraitDef}.
+ */
+public class SortRemoveConstantKeys extends RelOptRule {
+  public static final SortRemoveConstantKeys INSTANCE = new SortRemoveConstantKeys();
+
+  private SortRemoveConstantKeys() {
+    super(
+        operand(Sort.class, any()),
+        "SortRemoveConstantKeys");
+  }
+
+  @Override public void onMatch(RelOptRuleCall call) {
+    final Sort sort = call.rel(0);
+    final RelMetadataQuery mq = call.getMetadataQuery();
+    final RelOptPredicateList predicates =
+            mq.getPulledUpPredicates(sort.getInput());
+    if (predicates == null) {
+      return;
+    }
+
+    final List<RelFieldCollation> collationsList = new ArrayList<>();
+    for (RexNode currentNode : sort.getChildExps()) {
+
+      assert currentNode instanceof RexInputRef;
+      RexInputRef ref = (RexInputRef) currentNode;
+
+      if (!predicates.constantMap.containsKey(ref)) {
+        collationsList.add(new RelFieldCollation(ref.getIndex()));
+      }
+    }
+
+    if (collationsList.size() == sort.collation.getFieldCollations().size()) {
+      return;
+    }
+
+    // No active collations. Remove the sort completely
+    if (collationsList.isEmpty() && sort.offset == null && sort.fetch == null) {
+      call.transformTo(sort.getInput());
+
```

Comment:
you can also set the original operator sort's importance to 0

Reply:
Not sure how I can set the importance. I thought the cleanest way is to transform the call to a sort with the constant attribute removed. Could you please elaborate on the idea you suggested?

Reply:
like this:
https://github.com/apache/calcite/blob/d407395e05b5b37e019a44f812aae04540e6f398/core/src/main/java/org/apache/calcite/rel/rules/AggregateValuesRule.java#L106

You can do this too for L81. because the new plan is always better.

Reply:
Done, thanks